File: typescript/leetcodehard/palindromePartitioningII.ts
/*
https://leetcode.com/problems/palindrome-partitioning-ii/

status - completed
*/

import { logp } from "../utils/logger";

function minCut(s: string): number {
  const n = s.length;
  const isPalindrome: boolean[][] = Array.from({ length: n }, () => new Array(n).fill(false));

  for (let end = 0; end < n; end++) {
    for (let start = 0; start <= end; start++) {
      if (s[start] === s[end] && (end - start <= 2 || isPalindrome[start + 1][end - 1])) {
        isPalindrome[start][end] = true;
      }
    }
  }

  const dp = new Array(n).fill(Infinity);
  for (let end = 0; end < n; end++) {
    if (isPalindrome[0][end]) {
      dp[end] = 0;
      continue;
    }
    for (let start = 1; start <= end; start++) {
      if (isPalindrome[start][end]) {
        dp[end] = Math.min(dp[end], dp[start - 1] + 1);
      }
    }
  }

  return dp[n - 1];
}

logp(`input "aab" expected 1 actual ${minCut("aab")}`);
logp(`input "a" expected 0 actual ${minCut("a")}`);
logp(`input "ab" expected 1 actual ${minCut("ab")}`);

File: python/leetcodehard/palindrome_partitioning_ii.py
"""
https://leetcode.com/problems/palindrome-partitioning-ii/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def minCut(self, s):
        n = len(s)
        is_palindrome = [[False] * n for _ in range(n)]

        for end in range(n):
            for start in range(end + 1):
                if s[start] == s[end] and (end - start <= 2 or is_palindrome[start + 1][end - 1]):
                    is_palindrome[start][end] = True

        cuts = [0] * n
        for end in range(n):
            if is_palindrome[0][end]:
                cuts[end] = 0
                continue

            best = end
            for start in range(1, end + 1):
                if is_palindrome[start][end]:
                    best = min(best, cuts[start - 1] + 1)
            cuts[end] = best

        return cuts[n - 1] if n else 0


if __name__ == "__main__":
    test_input_1 = "aab"
    output = Solution().minCut(test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = "a"
    output = Solution().minCut(test_input_2)
    Solution().log.info("Output is %s", output)
